Lib/test/test_shelve.py

```diff
@@ -165,6 +167,22 @@ def test_default_protocol(self):
         with shelve.Shelf({}) as s:
             self.assertEqual(s._protocol, pickle.DEFAULT_PROTOCOL)
 
+    def test_custom_loads_and_dumps(self):
+        def custom_dumps(obj, protocol=None):
+            return bytes(f"{type(obj).__name__}", 'utf-8')
+
+        def custom_loads(data):
+            value = BytesIO(data).read()
+            return locate(value.decode("utf-8"))
+
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        with shelve.open(self.fn, custom_dumps=custom_dumps, custom_loads=custom_loads) as s:
+            num = 1
+            s['number'] = num
+            self.assertEqual(s['number'], type(num))
```

Comment:
Let's add test cases where dumps/loads/loads+dumps are passed individually with the proper use cases (also some negative tests that ensures we get a nice error when something is wrong)